### PR TITLE
RSCBC-1: Implement HTTP Crate

### PR DESCRIPTION
**Pull request by Matt-Woz**
Changes
---------
- Created HTTP Client, wrapping reqwest
- Created custom JSON Row parser + stream

### CODE REVIEW

File: sdk/couchbase-core/src/httpx/json_row_stream.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+where
+    S: Stream<Item = Result<Bytes, reqwest::Error>> + Unpin,
+{
+    type Item = HttpxResult<Vec<u8>>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<HttpxResult<Vec<u8>>>> {
+        let this = self.get_mut();
+        loop {
+            match this.state {
+                State::Collecting => match this.parser.next() {
```
Comment (chvck) [inline sdk/couchbase-core/src/httpx/json_row_stream.rs:56]: Nice! I think a few comments explaining each case here would be v. helpful.

File: sdk/couchbase-core/src/httpx/raw_json_row_streamer.rs [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+#[cfg(test)]
+mod tests {
+    use crate::httpx::base::{Body, Method, RequestBuilder};
+    use crate::httpx::client::Client;
+    use crate::httpx::json_row_stream::JsonRowStream;
+    use crate::httpx::raw_json_row_streamer::RawJsonRowStreamer;
+    use serde_json::{json, Value};
+
+    #[tokio::test]
+    async fn stream_some_query_rows() {
```
Comment (chvck) [inline sdk/couchbase-core/src/httpx/raw_json_row_streamer.rs:214]: Same comment as the other test.

**Comment by chvck** [comment]
This is great. It might be a good idea to move the integration tests to the top level `tests` crate and think about how we can add an abstraction layer between us and reqwest. That'd both decouple us a bit and let us unit test more easily.

File: sdk/couchbase-core/src/httpx/base.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+#[derive(Debug)]
+pub(crate) struct Body {
+    pub inner: reqwest::Body,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) struct OnBehalfOfInfo {
+    username: String,
```
[the remark this answers is not in the document]
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/base.rs]: Yep - done

File: sdk/couchbase-core/src/httpx/base.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    password: String,
+    domain: String,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) struct RequestBuilder {
```
[the remark this answers is not in the document]
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/base.rs]: Didn't know we had that! Done

File: sdk/couchbase-core/src/httpx/base.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            user_agent,
+            endpoint,
+            basic_auth_user,
+            basic_auth_pass,
+        }
+    }
+
+    pub(crate) fn new_request(
+        &self,
+        client: Client,
```
[the remark this answers is not in the document]
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/base.rs]: Yes it is a bit weird, reqwest recommends to use the RequestBuilder, which can be only accessed via a Client, and then running `RequestBuilder.send`, but I refactored to make it more intuitive

File: sdk/couchbase-core/src/httpx/client.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        match req.inner.send().await {
+            Ok(response) => Ok(Response { inner: response }),
+            Err(err) => {
+                if err.is_connect() {
+                    Err(Connect {
+                        msg: err.to_string(),
+                    }
+                    .into())
+                } else {
+                    Err(Generic {
```
[the remark this answers is not in the document]
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/client.rs:92]: Ack

File: sdk/couchbase-core/src/httpx/json_block_read.rs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        compat_version: String,
+        #[serde(alias = "clusterUUID")]
+        cluster_uuid: String,
+        #[serde(alias = "isBalanced")]
+        is_balanced: bool,
+        orchestrator: String,
+    }
+
+    #[tokio::test]
+    async fn get_cluster_info() {
```
[the remark this answers is not in the document]
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/json_block_read.rs]: I've moved the integration tests into tests/

File: sdk/couchbase-core/src/httpx/json_row_stream.rs (hunk 1 above)
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/json_row_stream.rs:56]: Done!

File: sdk/couchbase-core/src/httpx/base.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    Put,
+    Delete,
+    Connect,
+    Options,
+    Trace,
+    Patch,
+}
+#[derive(Debug)]
+pub(crate) struct Request {
+    pub inner: reqwest::RequestBuilder,
```
[the remark this answers is not in the document]
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/base.rs]: Yes agreed, I refactored to make it a lot more generic

File: sdk/couchbase-core/src/httpx/base.rs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    pub uri: String,
+    pub basic_auth_user: String,
+    pub basic_auth_pass: String,
+    pub method: http::Method,
+    pub content_type: String,
+    pub on_behalf_of: Option<OnBehalfOfInfo>,
+    pub body: Option<String>,
+}
+
+#[derive(PartialEq, Eq, Debug, TypedBuilder)]
```
Comment (chvck) [inline sdk/couchbase-core/src/httpx/base.rs:22]: All of the request type structs probably need to have #[non_exhaustive] on them.

File: sdk/couchbase-core/src/httpx/client.rs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+#[derive(Debug, Clone)]
+pub struct Client<C> {
+    pub inner: C,
+}
+
+#[async_trait]
+pub trait ClientProvider {
+    type R: ResponseProvider;
+    fn new(/*tlsConfig*/) -> HttpxResult<Self> where Self: Sized;
```
Comment (chvck) [inline sdk/couchbase-core/src/httpx/client.rs:18]: Could we file a rscbc for adding tls to http please?

**Review by chvck — COMMENTED** [review]
This is looking good to me, just the minor thing about #[non_exhaustive] so that expanding request blocks aren't breaking changes in the future. In the future we might want to think about using hyper but this shields any upstream crates from that too.

File: sdk/couchbase-core/src/httpx/base.rs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+use http::{StatusCode};
+use serde::de::DeserializeOwned;
+use typed_builder::TypedBuilder;
+use crate::httpx::error::ErrorKind::Generic;
+use crate::httpx::error::{Result as HttpxResult};
+
+#[derive(Debug, TypedBuilder)]
+#[builder(field_defaults(default, setter(into)))]
+#[non_exhaustive]
+pub struct Request {
```
Comment (chvck) [inline sdk/couchbase-core/src/httpx/base.rs:13]: Sorry, I missed this before. Most of these fields in these various structs should probably be Option<> if they are actually optional.

File: sdk/couchbase-core/src/httpx/client.rs (hunk 10 above)
Comment (Matt-Woz) [reply sdk/couchbase-core/src/httpx/client.rs]: Created RSCBC-44 👍

**Review by chvck — APPROVED** [review]

**Pull request merged**